src/runtime-rs/crates/hypervisor/src/device/driver/vfio.rs

```diff
@@ -22,56 +73,533 @@ fn override_driver(bdf: &str, driver: &str) -> Result<()> {
     Ok(())
 }
 
-const SYS_PCI_DEVICES_PATH: &str = "/sys/bus/pci/devices";
-const PCI_DRIVER_PROBE: &str = "/sys/bus/pci/drivers_probe";
-const VFIO_NEW_ID_PATH: &str = "/sys/bus/pci/drivers/vfio-pci/new_id";
-const VFIO_UNBIND_PATH: &str = "/sys/bus/pci/drivers/vfio-pci/unbind";
-
-pub const VFIO_PCI: &str = "vfio-pci";
-
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default, PartialEq)]
 pub enum VfioBusMode {
-    PCI,
+    #[default]
     MMIO,
+    PCI,
 }
 
 impl VfioBusMode {
-    pub fn new(mode: &str) -> Result<Self> {
-        Ok(match mode {
-            "mmio" => VfioBusMode::MMIO,
-            _ => VfioBusMode::PCI,
-        })
+    pub fn new(mode: &str) -> Self {
+        match mode {
+            "c" | "pci" => VfioBusMode::PCI,
+            _ => VfioBusMode::MMIO,
+        }
+    }
+
+    pub fn revert(mode: VfioBusMode) -> String {
+        match mode {
+            VfioBusMode::PCI => "pci".to_owned(),
+            VfioBusMode::MMIO => "mmio".to_owned(),
+        }
+    }
+
+    pub fn driver_type(mode: &str) -> &str {
+        match mode {
+            "c" | "pci" => DRIVER_VFIO_PCI_TYPE,
+            _ => DRIVER_MMIO_BLK_TYPE,
+        }
     }
 }
 
-#[derive(Debug, Clone)]
-pub struct VfioConfig {
+#[derive(Clone, Debug, Default)]
+pub enum VfioDeviceType {
+    // error type of VFIO device
+    Error,
+
+    // normal VFIO device type
+    #[default]
+    Normal,
+
+    // mediated VFIO device type
+    Mediated,
+}
+
+// represents a PCI device's device id and vendor id
+// DeVendor: (device, vendor)
+#[derive(Clone, Debug)]
+pub struct DeVendor(String, String);
+
+impl DeVendor {
+    pub fn new(vendor_id: &str, dev_id: &str) -> Self {
+        DeVendor(dev_id.to_owned(), vendor_id.to_owned())
+    }
+
+    pub fn get_device_vendor_id(&self) -> Result<u32> {
+        // default value is 0 when vendor_id or device_id is empty
```

Comment:
Maybe the line from 130 -147 can move the `DeVendor::new()`, the function get_device_vendor_id may call multi-times there is no need to parse each time.

Reply:
the new() will be remove. And I have a deep think that, the get_device_vendor_id will be keeped here for other cases.